### ISSUE TITLE
Installation issue: ncurses

### ISSUE THREAD

**Issue by ax3l** [LLNL/spack#33696]
### Steps to reproduce the issue

```console
$ spack spec -I ncurses
Input spec
--------------------------------
 -   ncurses

Concretized
--------------------------------
[+]  ncurses@6.3%gcc@10.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-skylake
[+]      ^pkgconf@1.8.0%gcc@10.3.0 arch=linux-ubuntu20.04-skylake
```
```
$ spack install ncurses
```

### Error message

<details><summary>Error message</summary><pre>
==> Error: Terminating after first install failure: AttributeError: 'str' object has no attribute 'propagate'

The 'ncurses' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
</pre></details>

### Information on your system

* **Spack:** 0.19.0.dev0 (68a5fe84a7f814af7f1c54f9701f33c5f4b2cd17)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04 on Azure Pipelines
* **Concretizer:** clingo

### Additional information

non-interactive execution

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by bernhardkaindl** [comment]
I cannot reproduce with ncurses, but reproduce a similar error with `spack -d install -v berkeleygw:
* **Spack:** 0.19.0.dev0 (0f0f5332b66562f182e9b24ea9f0fb650b7f3332)
* **Spack:** 0.20.0.dev0 (2f057d729da61e8c12828df44377f0a93fed820f)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Platform:** linux-ubuntu20.04-neoverse_n1
* **Concretizer:** clingo
After the install phase:
```
==> [2022-11-19-12:53:24.287421] Installing /home/bkaindl/spack/opt/spack/[padded-to-255-chars]/linux-ubuntu20.04-neoverse_n1/gcc-10.3.0/hdf5-1.12.2-ls2ucdsjatdseq2peyzvnkkksstzvemf/.spack/repos/builtin/packages/hdf5 to /home/bkaindl/spack/opt/spack/[padded-to-255-chars]/linux-ubuntu20.04-neoverse_n1/gcc-10.3.0/berkeleygw-3.0.1-fmnur2y47ch3ljvzhpejpmnschlmzexr/.spack/repos/builtin/packages/hdf5
==> [2022-11-19-12:53:24.289813] Reading config from file /home/bkaindl/spack/etc/spack/defaults/modules.yaml
==> [2022-11-19-12:53:24.302405] Reading config from file /home/bkaindl/spack/etc/spack/defaults/linux/modules.yaml
==> [2022-11-19-12:53:24.303965] Skipping nonexistent config path /etc/spack/modules.yaml
==> [2022-11-19-12:53:24.304063] Skipping nonexistent config path /etc/spack/linux/modules.yaml
==> [2022-11-19-12:53:24.304152] Skipping nonexistent config path /home/bkaindl/spack/etc/spack/modules.yaml
==> [2022-11-19-12:53:24.304240] Skipping nonexistent config path /home/bkaindl/spack/etc/spack/linux/modules.yaml
==> [2022-11-19-12:53:24.304316] Skipping nonexistent config path /home/bkaindl/.spack/modules.yaml
==> [2022-11-19-12:53:24.304386] Skipping nonexistent config path /home/bkaindl/.spack/linux/modules.yaml
==> [2022-11-19-12:53:24.330024] Error: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
See build log for details:
  /tmp/bkaindl/spack-stage/spack-stage-berkeleygw-3.0.1-fmnur2y47ch3ljvzhpejpmnschlmzexr/spack-build-out.txt

Traceback (most recent call last):
  File "/home/bkaindl/spack/lib/spack/spack/build_environment.py", line 1067, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/home/bkaindl/spack/lib/spack/spack/installer.py", line 2081, in build_process
    return installer.run()
  File "/home/bkaindl/spack/lib/spack/spack/installer.py", line 1953, in run
    spack.hooks.post_install(self.pkg.spec)
  File "/home/bkaindl/spack/lib/spack/spack/hooks/__init__.py", line 74, in __call__
    hook(*args, **kwargs)
  File "/home/bkaindl/spack/lib/spack/spack/hooks/module_file_generation.py", line 48, in post_install
    _for_each_enabled(spec, "write")
  File "/home/bkaindl/spack/lib/spack/spack/hooks/module_file_generation.py", line 32, in _for_each_enabled
    getattr(generator, method_name)()
  File "/home/bkaindl/spack/lib/spack/spack/modules/common.py", line 847, in write
    if self.conf.excluded:
  File "/home/bkaindl/spack/lib/spack/spack/modules/common.py", line 519, in excluded
    installed_implicitly = not spec._installed_explicitly()
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 4615, in _installed_explicitly
    record = spack.store.db.get_record(self)
  File "/home/bkaindl/spack/lib/spack/spack/database.py", line 122, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/bkaindl/spack/lib/spack/spack/database.py", line 1203, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/home/bkaindl/spack/lib/spack/spack/database.py", line 1198, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 4597, in __str__
    spec_str = " ^".join(d.format() for d in sorted_nodes)
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 4597, in <genexpr>
    spec_str = " ^".join(d.format() for d in sorted_nodes)
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 4328, in format
    write_attribute(self, attribute, color)
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 4314, in write_attribute
    write(sig + morph(spec, str(current)), col)
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 875, in __str__
    + " ".join(
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 877, in <genexpr>
    + ('=="' if True in [f.propagate for f in self[key]] else '="')
  File "/home/bkaindl/spack/lib/spack/spack/spec.py", line 877, in <listcomp>
    + ('=="' if True in [f.propagate for f in self[key]] else '="')
AttributeError: 'str' object has no attribute 'propagate'
==> [2022-11-19-12:53:24.330709] Flagging berkeleygw-3.0.1-fmnur2y47ch3ljvzhpejpmnschlmzexr as failed: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
==> [2022-11-19-12:53:24.377353] InstallError: Terminating after first install failure: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
==> [2022-11-19-12:53:24.377502] Error: Terminating after first install failure: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
Traceback (most recent call last):
  File "/home/bkaindl/spack/lib/spack/spack/installer.py", line 1748, in install
    self._install_task(task)
  File "/home/bkaindl/spack/lib/spack/spack/installer.py", line 1281, in _install_task
    spack.package_base.PackageBase._verbose = spack.build_environment.start_build_process(
  File "/home/bkaindl/spack/lib/spack/spack/build_environment.py", line 1231, in start_build_process
    raise child_result
spack.build_environment.ChildError: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bkaindl/spack/lib/spack/spack/main.py", line 1000, in main
    return _main(argv)
  File "/home/bkaindl/spack/lib/spack/spack/main.py", line 955, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/bkaindl/spack/lib/spack/spack/main.py", line 983, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/bkaindl/spack/lib/spack/spack/main.py", line 632, in _invoke_command
    return_val = command(parser, args)
  File "/home/bkaindl/spack/lib/spack/spack/cmd/install.py", line 549, in install
    install_specs(zip(abstract_specs, concrete_specs), install_kwargs, args)
  File "/home/bkaindl/spack/lib/spack/spack/cmd/install.py", line 270, in install_specs
    install_specs_outside_environment(specs, install_kwargs)
  File "/home/bkaindl/spack/lib/spack/spack/cmd/install.py", line 346, in install_specs_outside_environment
    builder.install()
  File "/home/bkaindl/spack/lib/spack/spack/installer.py", line 1795, in install
    raise InstallError("{0}: {1}".format(fail_fast_err, str(exc)), pkg=pkg)
spack.installer.InstallError: Terminating after first install failure: AttributeError: 'str' object has no attribute 'propagate'

The 'berkeleygw' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure
```
@ax3l: Would you also post the command output with run with `spack -d` so you get all debug messages and stack backtraces?

**Comment by bernhardkaindl** [comment]
@ax3l: I could reproduce it not with `ncurses` but with `berkeleygw`.

I bisected where the Error occurred first: It appears since https://github.com/spack/spack/pull/29761 (Enable/Disable Propagation for Compiler Flags & Variants) was merged.

I sent I pull request to fix this issue, see #34019, please have a try!

**Comment by bernhardkaindl** [comment]
@ax3l Nevermind, I found out that the ncurses install error was fixed by #33682.

The install error after berkelygw was triggered by the recipe doing `spec.compiler_flags["fflags"].append(...)`
